### DISCUSSION TITLE
How do I add inference results to a .txt file?

### DISCUSSION THREAD

**Discussion by yuvi-anand** [ultralytics/yolov5#8527] [Q&A] [answered]
I'm trying to add the results of my inference into a .txt file but I couldn't find a way to do that. Can anyone help?

**Comment by glenn-jocher** [comment, marked as answer]
@yuvi-anand you can use the --save-txt argument:
```
python detect.py --save-txt
```

**Reply by yuvi-anand** [reply]
Thank you so much for your help!

**Reply by yuvi-anand** [reply]
I had one more question, if my code looked something like this
<img width="633" alt="Screen Shot 2022-07-11 at 3 49 45 PM" src="https://user-images.githubusercontent.com/108440846/178371717-34468c33-01bb-4916-b6a9-b6b01b27f408.png">
how would I do the same thing (saving results to a text file) if I was not using detect.py?

**Reply by glenn-jocher** [reply]
@yuvi-anand 👋 Hello! Thanks for asking about **handling inference results**. YOLOv5 🚀 [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36) models allow for simple model loading and inference in a pure python environment without using `detect.py`.

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the YOLOv5 'small' model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints). Custom models can also be loaded, including custom trained PyTorch models and their [exported](https://github.com/ultralytics/yolov5/issues/251) variants, i.e. ONNX, TensorRT, TensorFlow, OpenVINO YOLOv5 models.

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # yolov5n - yolov5x6 official model
#                                            'custom', 'path/to/best.pt')  # custom model

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
results.xyxy[0]  # im predictions (tensor)

results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

results.pandas().xyxy[0].value_counts('name')  # class counts (pandas)
# person    2
# tie       1
```
<img width="550" src="https://user-images.githubusercontent.com/26833433/149679662-b2021c78-59b2-4f52-b768-1e5161537c85.png">

See [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36) for details.

Good luck 🍀 and let us know if you have any other questions!

**Reply by yuvi-anand** [reply]
Hi, I figured out how to make detect.py work on my directory, and I did use the --save-txt command. But I was wondering three things about this. One, how do I make it give the actual results (2 cars, 4 trucks, 1 bench) in the text file instead of the numbers that it returns? Two, if I'm running detect.py --save-txt on a directory, the results save in individual text files for each image. How do I make them save into a singular text file? And three, is it possible to format that text file? Thank you for all of your help, you have been a lifesaver!

**Reply by glenn-jocher** [reply]
@yuvi-anand I'd recommend the PyTorch Hub example I showed above for customizing inference output.
